typescript/cli/src/send/transfer.ts

```diff
@@ -179,74 +130,77 @@ async function executeDelivery({
   const provider = multiProvider.getProvider(origin);
   const connectedSigner = signer.connect(provider);
 
-  // TODO replace all code below with WarpCore
-  // https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3259
+  const warpCore = WarpCore.FromConfig(
+    MultiProtocolProvider.fromMultiProvider(multiProvider),
+    warpCoreConfig,
+  );
 
-  if (tokenType === TokenType.collateral) {
-    const wrappedToken = await getWrappedToken(routerAddress, provider);
-    const token = ERC20__factory.connect(wrappedToken, connectedSigner);
-    const approval = await token.allowance(signerAddress, routerAddress);
-    if (approval.lt(wei)) {
-      const approveTx = await token.approve(routerAddress, wei);
-      await approveTx.wait();
+  if (!routerAddress) {
+    const tokensForRoute = warpCore.getTokensForRoute(origin, destination);
+    if (tokensForRoute.length === 0) {
+      logRed(`No Warp Routes found from ${origin} to ${destination}`);
+      return;
     }
+
+    routerAddress = (await select({
+      message: `Select router address`,
+      choices: [
+        ...tokensForRoute.map((t) => ({
+          value: t.addressOrDenom,
+        })),
+      ],
+      pageSize: 10,
+    })) as string;
   }
 
-  let adapter: IHypTokenAdapter<PopulatedTransaction>;
-  const multiProtocolProvider =
-    MultiProtocolProvider.fromMultiProvider(multiProvider);
-  if (tokenType === TokenType.native) {
-    adapter = new EvmHypNativeAdapter(origin, multiProtocolProvider, {
-      token: routerAddress,
-    });
-  } else if (tokenType === TokenType.collateral) {
-    adapter = new EvmHypCollateralAdapter(origin, multiProtocolProvider, {
-      token: routerAddress,
-    });
-  } else {
-    adapter = new EvmHypSyntheticAdapter(origin, multiProtocolProvider, {
-      token: routerAddress,
-    });
+  const token = warpCore.findToken(origin, routerAddress);
+  if (!token) {
+    logRed(
+      `No Warp Routes found from ${origin} to ${destination} with router address ${routerAddress}`,
+    );
+    return;
+  }
+
+  const senderAddress = await signer.getAddress();
+  const errors = await warpCore.validateTransfer({
+    originTokenAmount: new TokenAmount(wei, token),
+    destination,
+    recipient: recipient ?? senderAddress,
+    sender: senderAddress,
+  });
+  if (errors) {
+    logRed('Unable to validate transfer');
```

Comment:
let's also log `errors` here, there shouldn't be any sensitive info in it

Reply:
thinking we should throw here as well